### PR TITLE
Credit contributor

### LINKED ISSUE #134: Contribution tool should append contributor's data to PR
As contributors should be credited for their contribution, the ability to specify the contributor when sending a document should be added.

### PR DESCRIPTION
**Pull request by martinratinaud**
Fixes #134

Here is a first implementation done on the contribution tool to achieve this.

When user contributes for the first time, s.he is asked to enter h.is.er email to take credit for the change.

If s.he does, he will become a co-author on the commit

(Text in the gif is not up to date though)
ota-contributor

@kissaki, I'm interested to know if you find the interface understandable enough.

### CODE REVIEW

**Comment by Kissaki** [comment]
How does it credit the authors with just the email? In Git a commit author is name + email.

Hard to see with the git immediately looping :P Looks like there is not only commit author but a co-authored-by footer too?

File: content/parts/en/contributor-form.mdx
Comment (MattiSG) [inline content/parts/en/contributor-form.mdx:23]:
```suggestion
## Sign your contribution

Open Terms Archive is an open source initiative that relies on contributors just like you 🤗 By default, your contributions will be anonymous. Signing your contributions with your email address will enable you to be notified when they are integrated or discussed and to be credited publicly for them.

<Input />
<UpdateButton>Sign your contribution with your email address</UpdateButton>

- While not readable anywhere on the web, this email address will be [publicly](https://stackoverflow.com/a/897604/594053) and irreversibly associated with your contribution.
- We will use this address only if it is necessary to contact you for integrating, crediting or otherwise handling this specific contribution.
- If you have an account on [GitHub](https://github.com), we recommend using an address linked to it so your account is mentioned.

<small>If you only want to contribute anonymously, you can <IgnoreButton>opt-out now</IgnoreButton>. You can always join the discussion later, but we will not be able to proactively contact you if we have questions.</small>
```

File: content/parts/fr/contributor-form.mdx
Comment (MattiSG) [inline content/parts/fr/contributor-form.mdx:1]: Let's work on this translation once we agree on an English version 🙂

File: locales/en/service.json
Comment (MattiSG) [inline locales/en/service.json:5]:
```suggestion
  "contributor.change": "Change signature",
```

File: locales/en/service.json
Comment (MattiSG) [inline locales/en/service.json:4]:
```suggestion
  "contributor.add": "Set signature",
```

File: locales/en/service.json
Comment (MattiSG) [inline locales/en/service.json:6]:
```suggestion
  "contributor.info": "Signing with <strong>{{email}}</strong>.",
```

File: src/modules/Common/data-components/ContributorForm.tsx
Comment (MattiSG) [inline src/modules/Common/data-components/ContributorForm.tsx:49]: > This is a very generic name.

```suggestion
              SignatureInput: () => (
```

File: src/modules/Common/data-components/ContributorForm.tsx
Comment (MattiSG) [inline src/modules/Common/data-components/ContributorForm.tsx:50]:
```suggestion
                <input required
```

File: src/modules/Common/data-components/ContributorForm.tsx
Comment (MattiSG) [inline src/modules/Common/data-components/ContributorForm.tsx:65]:
```suggestion
              OptOutButton: ({ children }: any) => <a onClick={onOptOut}>{children}</a>,
```

File: src/modules/Common/managers/ServiceManager.ts
Comment (MattiSG) [inline src/modules/Common/managers/ServiceManager.ts:68]:
```suggestion
Co-authored-by: Open Terms Archive Contributor <${this.authorEmail}>`;
```

**Review by MattiSG — COMMENTED** [review]
Nice, thanks!

I made some suggestions for copywriting, with the following aims:

1. Decreasing emphasis on “credit”, which is not an intrinsic motivator for all contributors and especially irrelevant for partners.
2. Clarifying usage and visibility of email address.
3. Switching to a vocabulary of “signature”, which I believe will be more widely understood.

I did not invest in translating until we agree on language 🙂

What would you think of simplifying and unifying the whole process by using a default email address of `anonymous.contributor@opentermsarchive.org`, for example? 🙂 Then, the UI would by default show “Contributing as anonymous contributor, change that”. This way:

1. No need for an opt-out setting.
2. No need to keep track of whether the modal has been shown or not.
3. We always have co-authors instead of having them only on some cases; in the future, if we manage to open PRs on behalf of the contributor or otherwise mention them, this will also be unified.
4. Both in Git and GitHub, we will be able to identify third-party anonymous contributions and separate them from explicit bot operations.

**Comment by martinratinaud** [comment]
Hi @Kissaki when the email is a GitHub email, no matter the name you pass in the `Co-authored-by`, it will associate your GitHub account.

Here is an example in a PR on our [sandbox declaration repo](https://github.com/OpenTermsArchive/sandbox-declarations/pull/165/commits)

**Comment by MattiSG** [comment]
> How does it credit the authors with just the email? In Git a commit author is name + email.

Indeed! In order to simplify the process and ask for only one piece of information, we just [use a generic name](https://github.com/OpenTermsArchive/contribution-tool/pull/135/files#diff-1f2b362829556c819bd660111bb493d93bbc3f8b6650549eb645fe331c4f98aeR68). If the author has a GitHub account associated with the email, it will take over whichever name has been written anyway.

As a contributor, please let us know if you would prefer to input both your name and email rather than having this implementation 🙂 

> Looks like there is not only commit author but a co-authored-by footer too?

Since we use the GitHub API for creating the commit, I understand from @martinratinaud that we cannot set the commit author ourselves: the identity is simply derived from the GitHub token. Instead of asking contributors for a GitHub token, which is cumbersome and few people have, we instead use the `Co-authored-by` text message to credit all users, no matter if they have a GitHub account or not.
In the future, we might offer the option of inputting a GitHub token (or signing in with GitHub), so that the PR opening itself is made on behalf of the user 🙂 but I wanted us to have a generic version covering most cases first.

**Comment by Kissaki** [comment]
Git commits have a commit author and a committer. I would expect you to be able to create commits as your bot user via the token, but declare a different commit author (name+email).

Co-authored-by is a ternary means, and allows associating multiple authors. It is not git-native, but is supported by various Git tools - including GitHub.

GitHub accounts may be identified and displayed on GitHub, but that is presentation specific to the GitHub interface and presentation. Coming from a git perspective I would expect and want to specify my name and email. Using a pseudo name is technically wrong, and breaks as soon as you leave GitHub.

We should also consider that GitHub offers email aliases for pseudo-anonymous commits (not publishing your own email). Entering that manually is not really feasible. A GitHub auth to retreive only the name + email would probably be best for regular GitHub users.

Using a name input defaulting to Anonymous and an email input defaulting to anonymous.contributor@opentermsarchive.org seems like a good first iteration/simple baseline to me.

Something else to consider is whether we want to simply accept unverified, unauthenticated email addresses. People could potentially specify email addresses they do not own. Authing via GitHub would remedy this, but only for people who do have a GitHub account.

---

I do not have access to the [sandbox declaration repo](https://github.com/OpenTermsArchive/sandbox-declarations/pull/165/commits)

**Comment by MattiSG** [comment]
> Given that the Author field exists, it seems unnecessary to use the co-authored-by field

I agree with you, as well as about remarks on how this setup is specific to GitHub. As [mentioned](https://github.com/OpenTermsArchive/contribution-tool/pull/135#issuecomment-1303077800) in my previous message, I had understood that we could not do this unless logging the user in through GitHub due to a limitation in the Octokit API.

However, looking at the `octokit.rest.repos.createOrUpdateFileContents` [documentation](https://octokit.github.io/rest.js/v19#repos-create-or-update-file-contents), it seems that one can pass a `author.name` and `author.email`. @martinratinaud is there thus any reason why we don't use those? 🙂

> unless you want to explicitly declare the bot/tool the author, and the user the co-author.

I believe it is relevant for the tool to be listed, since authors might have intended something different than what the tool created on their behalf. In that case, however, setting the `committer` information to the bot would seem the most relevant to me.

> We should also consider that GitHub offers email aliases for pseudo-anonymous commits (not publishing your own email). Entering that manually is not really feasible.

GitHub users can find their pseudonymous email address in their account settings (“emails” section), or [derive them](https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/setting-your-commit-email-address#about-commit-email-addresses) from their [user id](https://caius.github.io/github_id/#mattisg) in the form `ID+USERNAME@users.noreply.github.com`. However, this is a bit cumbersome to find.

We could use the GitHub API to retrieve the noreply email from the user ID, potentially through authentication, but I don't believe the cost-to-benefit ratio of that solution is good.

> Something else to consider is whether we want to simply accept unverified, unauthenticated email addresses.

At this stage, I believe this risk is acceptable. The impersonation risk is not significantly higher than through any other Git commit means. An attacker could forge a commit crediting an impersonated author name and email and open a pull request for it. The only difference is that here, we cannot identify the GitHub user that opened the pull request. However:

1. We have means to contact the credited author in case of doubts emerging from their authorship (attack scenario mitigated: make an offensive contribution, attribute it to target).
2. If the tool is marked as committer, authors have plausible deniability of their actions (attack scenario mitigated: track documents of a non-cooperative service provider, attribute them to employee of service provider).

> Authing via GitHub would remedy this, but only for people who do have a GitHub account.

I would prefer that we refrain from adding GitHub authentication at this stage, for two reasons:

1. I understand that setting up OAuth would require us to operate a dedicated database, while the contribution tool does not currently store any data.
2. The vendor-lock-in to GitHub is already too strong. Let's avoid adding more.

- - - -

> I do not have access to the [sandbox declaration repo](https://github.com/OpenTermsArchive/sandbox-declarations/pull/165/commits)

Unless other @OpenTermsArchive/core devs believe it would be more useful to open it, I believe it is important that this sandbox remains private to @OpenTermsArchive/core. This repository has been set up precisely to offer an opportunity to try out unstable features and track documents without risking being confused for production data. The current level of discoverability of production repositories is low, and `sandbox` and `contrib` could be very difficult to distinguish, for example.

- - - -

Based on this conversation, @martinratinaud what would you think of the following:

- Adding a Name field.
- Setting it by default to `Anonymous Contributor` and the email field to `anonymous@contribute.opentermsarchive.org`.
- Removing the opt-out process.
- Implementing `author` (to the provided name and email) and `committer` (to @OTA-Bot) in the call to `octokit.rest.repos.createOrUpdateFileContents`.

? 🙂

**Comment by martinratinaud** [comment]
Thanks a lot @Kissaki and @MattiSG for this suggestions, which I all agree with.
I will implement what's described.

**Comment by martinratinaud** [comment]
Here is how the new implementation looks like
ota-contributor-2

If you are fine with it @MattiSG please review the copy and I can merge it tomorrow

File: content/parts/en/contributor-form.mdx
Comment (MattiSG) [inline content/parts/en/contributor-form.mdx:3]:
```suggestion
Open Terms Archive is an open source initiative that relies on contributors just like you 🤗
```

File: content/parts/en/contributor-form.mdx
Comment (MattiSG) [inline content/parts/en/contributor-form.mdx:5]:
```suggestion
By default, your contributions will be anonymous. Signing your contributions will enable you to be notified when they are integrated or discussed and to be credited publicly for them.
```

File: content/parts/en/contributor-form.mdx
Comment (MattiSG) [inline content/parts/en/contributor-form.mdx:18]:
```suggestion
  the discussion later, but maintainers will not be able to proactively contact you if they have questions.
```

File: content/parts/fr/contributor-form.mdx
Comment (MattiSG) [inline content/parts/fr/contributor-form.mdx:1]:
```suggestion
## Signez votre contribution
```

File: src/modules/Common/managers/ServiceManager.ts
Comment (MattiSG) [inline src/modules/Common/managers/ServiceManager.ts:83]: This duplicates the [default values in `ContributorForm.tsx`](https://github.com/OpenTermsArchive/contribution-tool/pull/135/files#diff-9d1c0d6f31651577385f1d5f8856df453321b1f830a10e0b0610e28961fddaf6R9) that are set with `DEFAULT_CONTRIBUTOR_NAME` and `DEFAULT_CONTRIBUTOR_EMAIL`.

File: src/modules/Common/managers/ServiceManager.ts
Comment (MattiSG) [inline src/modules/Common/managers/ServiceManager.ts:164]: Is this really expected? It seems surprising that the `message` would be the same as the `title`.

File: src/pages/api/services/index.ts
Comment (MattiSG) [inline src/pages/api/services/index.ts:189]: > Is there no logging facility in this codebase? Using `console.error` seems more like a debugging instruction 🙂

File: src/pages/service.tsx
Comment (MattiSG) [inline src/pages/service.tsx:190]: > So in the end, we are still blocking on the first contribution in the given browser? I am not sure this is a satisfactory experience.

File: content/parts/fr/contributor-form.mdx
Comment (MattiSG) [reply content/parts/fr/contributor-form.mdx:1]: This translation is outdated.

File: locales/fr/service.json
Comment (MattiSG) [inline locales/fr/service.json:6]: This translation is outdated.

**Review by MattiSG — CHANGES_REQUESTED** [review]

File: src/pages/service.tsx
Comment (clementbiron) [reply src/pages/service.tsx:190]: Agree, we should not force user to view the contributor modal.

**Review by clementbiron — CHANGES_REQUESTED** [review]
As Matti said, french translations of `contributor-form.mdx` file is outdated.

I pushed some UI improvements.

File: locales/en/service.json
Comment (MattiSG) [inline locales/en/service.json:4]: Is this string still used anywhere?

File: src/pages/service.tsx
Comment (martinratinaud) [reply src/pages/service.tsx:190]: I'm pretty sure if we do not do this, the little link below the button saying "Set signature" may not be enough to produce this kind of engagement.

File: src/modules/Common/managers/ServiceManager.ts
Comment (martinratinaud) [reply src/modules/Common/managers/ServiceManager.ts:164]: As message is "Add <service> <documentType>" I believe it is explanatory enough

In order to be coherent with other commits, I would have to use "Add <service> <documentType> declaration" which I can do.

File: src/modules/Common/managers/ServiceManager.ts
Comment (martinratinaud) [reply src/modules/Common/managers/ServiceManager.ts:83]: Right,
as this value is used on both frontend and server, i will have to add it to the `publicRuntime` next config.

File: src/pages/api/services/index.ts
Comment (martinratinaud) [reply src/pages/api/services/index.ts:189]: There is indeed none. I didn't feel I needed one.

File: src/pages/service.tsx
Comment (MattiSG) [reply src/pages/service.tsx:190]: I'm pretty sure if we ask for name and email, we will have increased drop rate from uncertain first-time contributors :wink: These are just assumptions against other assumptions… only user testing could tell us that.

What I do know is that in the main use case we have for the contribution tool now, i.e. demonstration to potential partners and then engaged partners, having an additional mandatory step when they first see or try it increases again the felt complexity of an already complex interface, and that once they start working with it on a daily basis it will be very easy to have them set their signature up.

Comment (MattiSG) [reply src/pages/service.tsx:190]: I would agree with you @martinratinaud if we were building a system targeting first-time, potential one-shot contributors. This is not the case at the moment 🙂

Comment (martinratinaud) [reply src/pages/service.tsx:190]: I agree with this point of view and will thus make the changes

File: content/parts/en/contributor-form.mdx
Comment (Kissaki) [inline content/parts/en/contributor-form.mdx:9]: > While not readable anywhere on the web

I don’t think that’s true? Or am I missing something?

The repository is hosted on GitHub, and GitHub displays the Git commit information like author and author email fields as well as the commit message.

File: content/parts/en/contributor-form.mdx
Comment (Kissaki) [inline content/parts/en/contributor-form.mdx:14]: I would drop the *only*

```suggestion
  If you want to contribute anonymously, you can leave the default email. You can always join the discussion later, but maintainers will not be able to proactively contact you if they have questions.
```

It simplifies the sentence. One may argue about the subjective superiority/implied value/nudging - but I think even without that the simpler sentence is worth it.

For us it does not really matter if a contribution is anonymous or credited.

Comment (Kissaki) [inline content/parts/en/contributor-form.mdx:14]: > You can always join the discussion later, but maintainers will not be able to proactively contact you if they have questions.

I assume this is mainly referring to joining a PR discussion, and maintainers contacting related to the PR before merging?

With the email use disclosure also listing “otherwise handling this specific contribution” I was also thinking of if there is relicensing and other copyright related requests/communication. Is that also considered a possibility, and if not, what does other handling refer to/enclose?

**Review by Kissaki — COMMENTED** [review]
I reviewed the text, for your consideration. I hope that’s ok.

File: content/parts/en/contributor-form.mdx
Comment (martinratinaud) [reply content/parts/en/contributor-form.mdx:9]: You're right on the fact that it can be read, but on the commit page through, I could not find any place mentioning the email.

Screen Shot 2022-11-14 at 11 36 33

It's available through an API call though.

Screen Shot 2022-11-14 at 11 37 59

Comment (MattiSG) [reply content/parts/en/contributor-form.mdx:9]: Indeed, thanks @martinratinaud for the examples 🙂 

Hence my phrasing: this is not available on the web, but is public. Anyone cloning the repo can find it. Not through a web search though, and this will not be indexed.

File: jest.config.js
Comment (MattiSG) [inline jest.config.js:4]: > These template comments do not seem useful here.

File: next.config.js
Comment (MattiSG) [inline next.config.js:12]:
```suggestion
    author: {
```

File: src/modules/Common/data-components/ContributorForm.tsx
Comment (MattiSG) [inline src/modules/Common/data-components/ContributorForm.tsx:30]: > I'm not sure of the relevance of prefixing this with `ota` considering that localStorage is scoped per domain.

**Review by MattiSG — APPROVED** [review]

**Review by MattiSG — APPROVED** [review]
Already approved 🙂  Thanks for having taken into account optional comments!

**Review by clementbiron — APPROVED** [review]

**Comment by martinratinaud** [comment]
Thanks a lot @Kissaki for your help and insights. The feature is live on the contribution website and we've been wanting for a long time :-).

**Pull request merged**
